nested-data-and-loops/nested-loops.md

```diff
@@ -0,0 +1,203 @@
+
+We then loop through every character in each word. For every character, we will run this set of instructions: we'll check if the character is in our `char_map` dictionary as a key already. If it is _not_ in the `char_map` dictionary already, then this is the first time we're seeing it, and it should go in the dictionary with a count of one. Otherwise, it's already in the dictionary, and we can increment the frequency count.
+
+Callout: Let's reword this logic!
+
+We can trace the logic going from "most inner code" to "most outer": We are touching `char_map` in only one place in the code: the `if...else` block. However, `char_map` gets modified for every character, in every word from our original list of words.
+
+Tracing the logic going from "most outer code" to "most inner": We are looping through every word, and then every character in every word, before asking if the character is in `char_map` .
+
+<!-- end callout -->
+
+The function ultimately returns `char_map` . The code outside of the function definition create an array of words in `colors` , and then passes it into the function. When we get the character frequency map back, we print it to the console.
+
+### Practice
+
+Read through this code and answer these questions.
+
+``` python
+def map_character_frequency(words):
+    char_map = {}
+    for word in words:
+        for character in word:
+            if character not in char_map.keys():
+                char_map[character] = 1
+            else:
+                char_map[character] += 1
+    return char_map
+
+colors = ["red", "orange", "yellow", "green"]
+
+colors_char_map = map_character_frequency(colors)
+print(colors_char_map)
+```
+
+How many loops are in the function `map_character_frequency()` ?
+
+* 1 for-loop, 1 while-loop
+* 2 for-loops, 0 while-loops
+* 0 for-loops, 2 while-loops
+* 2 for-loops, 0 while-loops, 1 if-else loop
+
+Which line of code best describes iterating over the list `words` ?
+
+* `for word in words:`
+* `for character in word:`
+* `if character in char_map.keys():`
+* `char_map[character] += 1`
+
+What is the local variable for each item in `words` ?
+
+* `char_map`
+* `word`
+* `words`
+* `character`
+
+Which line of code best describes iterating over each character in `word` ?
+
+* `for word in words:`
+* `for character in word:`
+* `if character in char_map.keys():`
+* `char_map[character] += 1`
+
+What is the local variable for each item in `word` ?
+
+* `char_map`
+* `word`
+* `words`
+* `character`
+
+### Tips
+
+* Use indentation to understand the nesting
+* Take time to understand the local variables in each loop
+* Use words, comments, and pseudocode, and use the phrase "for each ..., we do ..." and "while ... is true, we do ..."
+
+## Visualizing Loops in Loops
+
+It can be a challenge to visualize how many loops run. To visualize nested loops, we can use a loop table. Let's use this code as our example.
```

Comment:
Rephrased to narrow perceived difficulty, and to focus on presenting the tool.
```suggestion
Sometimes it can be challenging to visualize how a particular loop runs. A _**loop table**_ is a tool we can use to help us visualize the operation of loops, both nested and unnested. Let's use our now-familiar character counting code as an example.
```